samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java

```diff
@@ -114,6 +100,11 @@ public boolean tryBecomeLeader() {
     String predecessor = children.get(index - 1);
     if (!predecessor.equals(currentSubscription)) {
       if (currentSubscription != null) {
+
+        // callback in case if the previous node gets deleted (when previous processor dies)
```

Comment:
We have discussed this before. It is not a bug simply because we have a recursive call, granted we can re-write it in an iterative style. 
We need this check and re-try because zkClient does not throw exceptions or fail when it tries to subscribe to a non-existent path. This is needed for a non-leader when it tries to subscribe to a predecessor node. If the node went away when the follower tried to subscribe to changes, it won't get any callbacks and will be orphaned. In order to avoid that scenario, we check with an "exists" of the predecessor node AFTER subscription. If it doesn't exist, we need to re-try leader election. 
I believe I added a comment around why we do this check.